File: tally-android/core/domain/src/main/java/com/tallytracker/domain/util/DataPortability.kt
package com.tallytracker.domain.util

import com.tallytracker.domain.model.Challenge
import com.tallytracker.domain.model.Entry
import com.tallytracker.domain.model.User
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * Handles data export and import for user data portability
 */
object DataPortability {
    
    private val json = Json {
        prettyPrint = true
        ignoreUnknownKeys = true
    }
    
    @Serializable
    data class ExportData(
        val exportDate: String,
        val version: String,
        val user: ExportUser?,
        val challenges: List<ExportChallenge>,
        val entries: List<ExportEntry>
    )
    
    @Serializable
    data class ExportUser(
        val id: String,
        val email: String?,
        val name: String?
    )
    
    @Serializable
    data class ExportChallenge(
        val id: String,
        val name: String,
        val target: Int,
        val color: String,
        val icon: String,
        val unit: String,
        val year: Int,
        val isPublic: Boolean
    )
    
    @Serializable
    data class ExportEntry(
        val id: String,
        val challengeId: String,
        val date: String,
        val count: Int,
        val note: String?
    )
    
    /**
     * Export all user data as JSON string
     */
    fun exportAllData(
        user: User?,
        challenges: List<Challenge>,
        entries: List<Entry>
    ): String {
        val exportData = ExportData(
            exportDate = LocalDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME),
            version = "1.0.0",
            user = user?.let { ExportUser(it.id, it.email, it.name) },
            challenges = challenges.map { c ->
                ExportChallenge(
                    id = c.id,
                    name = c.name,
                    target = c.target,
                    color = c.color,
                    icon = c.icon,
                    unit = c.unit,
                    year = c.year,
                    isPublic = c.isPublic
                )
            },
            entries = entries.map { e ->
                ExportEntry(
                    id = e.id,
                    challengeId = e.challengeId,
                    date = e.date,
                    count = e.count,
                    note = e.note
                )
            }
        )
        
        return json.encodeToString(exportData)
    }
    
    /**
     * Parse import data from JSON string
     */
    fun parseImportData(jsonString: String): ExportData {
        return json.decodeFromString(jsonString)
    }
}

File: tally-ios/Tally/Tally/Sources/Utilities/DataPortability.swift
import Foundation

/// Handles data export and import for user data portability
actor DataPortability {
    static let shared = DataPortability()
    
    private init() {}
    
    // MARK: - Export
    
    struct ExportData: Codable {
        let exportDate: Date
        let version: String
        let user: AuthUser?
        let challenges: [ChallengeResponse]
        let entries: [EntryResponse]
    }
    
    /// Export all user data as JSON string
    func exportAllData() async throws -> String {
        let challenges = try await APIClient.shared.getChallenges()
        var allEntries: [EntryResponse] = []
        
        for challenge in challenges {
            let entries = try await APIClient.shared.getEntries(challengeId: challenge.id)
            allEntries.append(contentsOf: entries)
        }
        
        let exportData = ExportData(
            exportDate: Date(),
            version: "1.0.0",
            user: AuthManager.shared.currentUser,
            challenges: challenges,
            entries: allEntries
        )
        
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        
        let data = try encoder.encode(exportData)
        guard let string = String(data: data, encoding: .utf8) else {
            throw DataPortabilityError.encodingFailed
        }
        
        return string
    }
    
    // MARK: - Import
    
    /// Import data from JSON string
    func importData(from jsonString: String) async throws {
        guard let data = jsonString.data(using: .utf8) else {
            throw DataPortabilityError.decodingFailed
        }
        
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        
        let importData = try decoder.decode(ExportData.self, from: data)
        
        // Import challenges
        for challenge in importData.challenges {
            _ = try await APIClient.shared.createChallenge(
                title: challenge.title,
                target: challenge.target,
                unit: challenge.unit,
                color: challenge.color,
                icon: challenge.icon,
                startDate: challenge.startDate,
                endDate: challenge.endDate,
                isPublic: challenge.isPublic
            )
        }
        
        // Note: Entries would need to be imported with their original challengeIds mapped
        // This is a simplified version - full implementation would need ID mapping
    }
}

enum DataPortabilityError: Error, LocalizedError {
    case encodingFailed
    case decodingFailed
    case importFailed(String)
    
    var errorDescription: String? {
        switch self {
        case .encodingFailed:
            return "Failed to encode data for export"
        case .decodingFailed:
            return "Failed to decode import data"
        case .importFailed(let reason):
            return "Import failed: \(reason)"
        }
    }
}
